barretenberg/cpp/src/barretenberg/vm2/constraining/relations/emit_notehash.test.cpp

```diff
@@ -0,0 +1,145 @@
+using simulation::NoteHashTreeCheckEvent;
+
+using testing::_;
+using testing::NiceMock;
+using testing::Return;
+
+using FF = AvmFlavorSettings::FF;
+using C = Column;
+using emit_notehash = bb::avm2::emit_notehash<FF>;
+using RawPoseidon2 = crypto::Poseidon2<crypto::Poseidon2Bn254ScalarFieldParams>;
+
+TEST(EmitNoteHashConstrainingTest, Positive)
+{
+    uint64_t prev_num_note_hashes_emitted = MAX_NOTE_HASHES_PER_TX - 1;
+    TestTraceContainer trace({ {
+        { C::execution_sel_execute_emit_notehash, 1 },
+        { C::execution_register_0_, /*note_hash=*/42 },
+        { C::execution_mem_tag_reg_0_, static_cast<uint8_t>(MemoryTag::FF) },
+        { C::execution_remaining_note_hashes_inv, FF(MAX_NOTE_HASHES_PER_TX - prev_num_note_hashes_emitted).invert() },
+        { C::execution_sel_write_note_hash, 1 },
+        { C::execution_sel_opcode_error, 0 },
+        { C::execution_subtrace_operation_id, AVM_EXEC_OP_ID_EMIT_NOTEHASH },
+        { C::execution_prev_note_hash_tree_size, 1 },
+        { C::execution_note_hash_tree_size, 2 },
+        { C::execution_prev_num_note_hashes_emitted, prev_num_note_hashes_emitted },
+        { C::execution_num_note_hashes_emitted, prev_num_note_hashes_emitted + 1 },
+    } });
+    check_relation<emit_notehash>(trace);
+}
+
+TEST(EmitNoteHashConstrainingTest, LimitReached)
+{
+    uint64_t prev_num_note_hashes_emitted = MAX_NOTE_HASHES_PER_TX;
+    TestTraceContainer trace({ {
+        { C::execution_sel_execute_emit_notehash, 1 },
+        { C::execution_register_0_, /*note_hash=*/42 },
+        { C::execution_mem_tag_reg_0_, static_cast<uint8_t>(MemoryTag::FF) },
+        { C::execution_remaining_note_hashes_inv, 0 },
+        { C::execution_sel_write_note_hash, 0 },
+        { C::execution_sel_opcode_error, 1 },
+        { C::execution_subtrace_operation_id, AVM_EXEC_OP_ID_EMIT_NOTEHASH },
+        { C::execution_prev_note_hash_tree_size, 1 },
+        { C::execution_note_hash_tree_size, 1 },
+        { C::execution_prev_num_note_hashes_emitted, prev_num_note_hashes_emitted },
+        { C::execution_num_note_hashes_emitted, prev_num_note_hashes_emitted },
+    } });
+    check_relation<emit_notehash>(trace);
+
+    // Negative test: sel_opcode_error must be on
+    trace.set(0, { { { C::execution_sel_opcode_error, 0 } } });
+    EXPECT_THROW_WITH_MESSAGE(check_relation<emit_notehash>(trace, emit_notehash::SR_EMIT_NOTEHASH_LIMIT_REACHED),
+                              "EMIT_NOTEHASH_LIMIT_REACHED");
+
+    // Negative test: tree size must be the same
+    trace.set(0, { { { C::execution_note_hash_tree_size, 2 } } });
+    EXPECT_THROW_WITH_MESSAGE(check_relation<emit_notehash>(trace, emit_notehash::SR_EMIT_NOTEHASH_TREE_SIZE_INCREASE),
+                              "EMIT_NOTEHASH_TREE_SIZE_INCREASE");
+
+    // Negative test: num note hashes emitted must be the same
+    trace.set(0, { { { C::execution_num_note_hashes_emitted, prev_num_note_hashes_emitted + 1 } } });
```

Comment:
when setting just one row prefer the method that's not taking in a span

trace.set(C::execution_num_note_hashes_emitted, /*row=*/0, prev_num_note_hashes_emitted + 1);